File: backend/studio/routes/prompt_favorites.py
"""
studio.routes.prompt_favorites — prompt history 즐겨찾기 CRUD.

엔드포인트 (prefix=/api/studio):
  - GET    /prompt-favorites             — 전체 또는 mode별 목록
  - POST   /prompt-favorites             — mode + 사용자 원문 prompt 저장(upsert)
  - DELETE /prompt-favorites/{favorite_id} — 즐겨찾기 해제
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from .. import history_db

router = APIRouter()


class PromptFavoriteBody(BaseModel):
    mode: str = Field(..., min_length=1)
    prompt: str = Field(..., min_length=1)


@router.get("/prompt-favorites")
async def list_prompt_favorites(mode: str | None = None):
    """프롬프트 즐겨찾기 목록 — updated_at 최신순."""
    safe_mode = (
        mode if mode in history_db.VALID_PROMPT_FAVORITE_MODES else None
    )
    items = await history_db.list_prompt_favorites(safe_mode)
    return {"items": items}


@router.post("/prompt-favorites")
async def create_prompt_favorite(body: PromptFavoriteBody):
    """사용자 원문 prompt 를 즐겨찾기에 저장. 같은 mode+prompt 는 갱신."""
    try:
        item = await history_db.upsert_prompt_favorite(body.mode, body.prompt)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    return {"item": item}


@router.delete("/prompt-favorites/{favorite_id}")
async def delete_prompt_favorite(favorite_id: str):
    """프롬프트 즐겨찾기 해제."""
    ok = await history_db.delete_prompt_favorite(favorite_id)
    if not ok:
        raise HTTPException(404, "favorite not found")
    return {"ok": True}

File: frontend/lib/api/prompt-favorites.ts
/**
 * lib/api/prompt-favorites.ts — PromptHistoryPeek 별 즐겨찾기 API.
 *
 * 저장 대상은 사용자가 입력한 원문 prompt 이며, AI 보강 후 실행 prompt 는
 * 생성 결과 history 의 upgradedPrompt 로만 유지한다.
 */

import { STUDIO_BASE, USE_MOCK, uid } from "./client";
import type { PromptFavorite, PromptFavoriteMode } from "./types";

export async function listPromptFavorites(opts?: {
  mode?: PromptFavoriteMode;
}): Promise<PromptFavorite[]> {
  if (USE_MOCK) return [];
  const q = new URLSearchParams();
  if (opts?.mode) q.set("mode", opts.mode);
  const suffix = q.toString() ? `?${q.toString()}` : "";
  const res = await fetch(`${STUDIO_BASE}/api/studio/prompt-favorites${suffix}`);
  if (!res.ok) throw new Error(`prompt favorites list failed: ${res.status}`);
  const data = (await res.json()) as { items: PromptFavorite[] };
  return data.items ?? [];
}

export async function createPromptFavorite(input: {
  mode: PromptFavoriteMode;
  prompt: string;
}): Promise<PromptFavorite> {
  const prompt = input.prompt.trim();
  if (!prompt) throw new Error("prompt required");

  if (USE_MOCK) {
    const now = Date.now();
    return {
      id: uid("fav"),
      mode: input.mode,
      prompt,
      promptHash: `mock-${now}`,
      createdAt: now,
      updatedAt: now,
    };
  }

  const res = await fetch(`${STUDIO_BASE}/api/studio/prompt-favorites`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ mode: input.mode, prompt }),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`prompt favorite create failed: ${res.status} ${text}`);
  }
  const data = (await res.json()) as { item: PromptFavorite };
  return data.item;
}

export async function deletePromptFavorite(id: string): Promise<boolean> {
  if (USE_MOCK) return true;
  const res = await fetch(
    `${STUDIO_BASE}/api/studio/prompt-favorites/${encodeURIComponent(id)}`,
    { method: "DELETE" },
  );
  if (res.status === 404) return false;
  if (!res.ok) throw new Error(`prompt favorite delete failed: ${res.status}`);
  return true;
}
